Sources/SwiftCompilerPlugin/PluginMacroExpansionContext.swift

```diff
@@ -0,0 +1,253 @@
+      node = Syntax(AttributeSyntax.parse(from: &parser))
+    }
+    do {
+      node = try OperatorTable.standardOperators.foldAll(node)
+    } catch let error {
+      // TODO: Error handling.
+      fatalError(error.localizedDescription)
+    }
+
+    let location = KnownSourceSyntax.Location(
+      offset: syntaxInfo.location.offset,
+      line: syntaxInfo.location.line,
+      column: syntaxInfo.location.column,
+      fileID: syntaxInfo.location.fileID,
+      fileName: syntaxInfo.location.fileName
+    )
+
+    knownSourceSyntax[node.id] = KnownSourceSyntax(node: node, location: location)
+
+    return node
+  }
+
+  /// Get position (file name + offset) of `node` in the known root nodes.
+  /// The root node of `node` must be one of the retured value from `add(_:)`.
+  func position(
+    of node: Syntax,
+    at kind: PositionInSyntaxNode
+  ) -> SourcePosition? {
+    guard let base = self.knownSourceSyntax[node.root.id] else {
+      return nil
+    }
+    let localPosition = node.position(at: kind)
+    let positionOffset = base.location.offset
+    return SourcePosition(
+      fileName: base.location.fileName,
+      utf8Offset: localPosition.advanced(by: positionOffset).utf8Offset
+    )
+  }
+
+  /// Get position range of `node` in the known root nodes.
+  /// The root node of `node` must be one of the retured value from `add(_:)`.
+  func range(
+    of node: Syntax,
+    from startKind: PositionInSyntaxNode = .afterLeadingTrivia,
+    to endKind: PositionInSyntaxNode = .beforeTrailingTrivia
+  ) -> SourceRange? {
+    guard let base = self.knownSourceSyntax[node.root.id] else {
+      return nil
+    }
+    let localStartPosition = node.position(at: startKind)
+    let localEndPosition = node.position(at: endKind)
+    assert(localStartPosition <= localEndPosition)
+
+    let positionOffset = base.location.offset
+
+    return SourceRange(
+      fileName: base.location.fileName,
+      startUTF8Offset: localStartPosition.advanced(by: positionOffset).utf8Offset,
+      endUTF8Offset: localEndPosition.advanced(by: positionOffset).utf8Offset
+    )
+  }
+
+  /// Get location of `node` in the known root nodes.
+  /// The root node of `node` must be one of the retured value from `add(_:)`.
+  func location(of node: Syntax, at kind: PositionInSyntaxNode, filePathMode: SourceLocationFilePathMode) -> SourceLocation? {
+    guard let base = self.knownSourceSyntax[node.root.id] else {
+      return nil
+    }
+    let file: String
+    switch filePathMode {
+    case .fileID: file = base.location.fileID
+    case .filePath: file = base.location.fileName
+    }
+
+    let localPosition = node.position(at: kind)
+    let localLocation = base.locationConverter.location(for: localPosition)
+
+    let positionOffset = base.location.offset
+    let lineOffset = base.location.line - 1
+    let columnOffset = localLocation.line == 1 ? base.location.column : 0
+
+    return SourceLocation(
+      line: localLocation.line! + lineOffset,
```

Comment:
I forgot why `SourceLocation` has an optional `line` and `column` but are we sure that these are non-nil here? If yes, could you add a doc comment explaining why we can assume they are non-nil here?

Reply:
`SourceLocation.line/column` are nil [only if the location converter was not supplied](https://github.com/apple/swift-syntax/blob/7fa6666bb8bde61ce3af2389a2731538091c2948/Sources/SwiftSyntax/SourceLocation.swift#L82-L91). And yes I am sure these cannot be nil, if the `offset` is out of the range, it's clamped to the range of the file.

And yes, I can add a comment.